Implement SequenceLocks functions for BIP 68

morcos opened the pull request:
SequenceLocks functions are used to evaluate sequence lock times or heights per BIP 68.

This is an alternate to #6312 

~~This PR still needs to update unit tests.~~
In addition it no longer reflects sequence locked information in wallet txs.  This can be added in later if desired, but I don't think its important now.

This code borrows heavily from #6312 and the work of @maaku, @btcdrak, @NicolasDorier, and @sipa 

(EDIT) **Important Note:**  This PR has changed the semantics of BIP 68 to always use MTP for comparison regardless of BIP 113.  I believe this makes more sense.  BIP 113 is still needed to change the semantics of nLockTime however.

morcos committed:
Implement SequenceLocks functions

SequenceLocks functions are used to evaluate sequence lock times or heights per BIP 68.

The majority of this code is copied from maaku in #6312
Further credit: btcdrak, sipa, NicolasDorier

NicolasDorier commented:
wow I like that, the changes are very small to review. I'll add the tests later tonight or tomorrow.

morcos commented:
@NicolasDorier oh good, i was mostly holding off on the tests because i wanted to see if people liked this approach.

What do you mean about IsFinalTx?  I think the changes to make that aware of MTP are already merged, so it wasn't necessary for me to change it further.

instagibbs commented on src/main.h:
```diff
@@ -341,7 +341,22 @@ bool IsFinalTx(const CTransaction &tx, int nBlockHeight, int64_t nBlockTime);
  */
 bool CheckFinalTx(const CTransaction &tx, int flags = -1);
```
Like the symmetry, but the name here is now a bit out-dated and possibly misleading since we have two types of finality.

morcos replied:
Agreed.  I was going for minimal changes to existing consensus code, but if people like I'd be happy to rename that.  Perhaps `LocktimeLock()` and `CheckLocktimeLock()`

I'd also like to take out the way CheckFinalTx is called with default -1 in the wallet code, but I could save that for another pull.

jtimon replied:
IsFinalTx is planned to be part of libconsensus,   CheckFinalTx is just a convenience function using globals. I don't  consider CheckFinalTx (nor CheckSequenceLocks) a "consensus  function" or "consensus code".

NicolasDorier commented:
I am trying another approach before building on yours, which would solve the removeForReorg and CNB problem without changing any code in #6312.

I removed my comment about IsFinalTx, I was surprised not seeing the flag passed to it. But it seems client are held responsible for calculating the right blocktime. This led to code duplication, #6312 fixed that.

I like the idea of having 2 methods : CheckLockTime and CheckSequenceLockTime though, but I wonder if it will really be useful since there will be no case where we want to call one and not the other.

morcos commented:
added unit tests

sdaftuar commented:
@NicolasDorier One reason I like the approach in this PR is that the sequence-lock check is done separately from the nLockTime check -- locktime can be evaluated for all transactions in a block at the time a block is received (eg in `ContextualCheckBlock`), whereas sequence locks can only be evaluated when all inputs heights' are available (eg during `ConnectBlock`).

From a design perspective, it seems to me that having the consensus checks operate in the smallest logical units ought to limit the need for future refactorings (which I think is a good design goal for consensus code).

NicolasDorier commented:
Ok, it makes sense. I'm kind of worried though about having the block of CheckLockTime responsible for calculating the blocktime each time, this is code duplication easy to get wrong. (I think that when I fixed it in #6312, there was indeed some parts which could break if the block evaluated was the genesis with MTP enabled)

NicolasDorier commented:
It does not solve the perf problem though, CheckSequenceLock need to be called during reorg. (Did : https://github.com/NicolasDorier/bitcoin/commit/f7f5055c3b40025c54550978b003e652ba1f64b7 it takes time for me to compile, I don't have a linux at hand, doing it by pushing/waiting travis)

morcos commented:
Here is another version of this that assumes BIP68 includes MTP already: https://github.com/morcos/bitcoin/commit/ba957d5407ecc3d1f320893f47cfc7181a0a0878

btcdrak commented:
@morcos Since we plan to roll out BIP113 and BIP68 together it really makes sense that BIP68 assumes MTP. I think that is a must and it simplifies things.

jtimon commented:
NACK This wastes a lot of previous review effort by not building on top of #6312.

sdaftuar committed:
Add RPC test exercising BIP68 (mempool only)

morcos commented:
OK I fixed my comment about -1 and switched to the version where BIP 68 is defined to use MTP.

If it is helpful for anyone to see what this PR looks like as a change from #6312, you can see that here:
https://github.com/maaku/bitcoin/compare/sequencenumbers...morcos:7184onorig6312

jtimon commented:
> If it is helpful for anyone to see what this PR looks like as a change from #6312, you can see that here: maaku@sequencenumbers...morcos:7184onorig6312

Thank you, it is certainly useful, at least for me.
Why not use that branch here directly (like @sdaftuar and @NicolasDorier are doing with their solutions) instead? Later maybe you can squash some of the commits if this replaces #6312 .
I don't understand why you insist in doing it in a new single commit with your name in it instead.

sdaftuar commented on src/primitives/transaction.h:
```diff
@@ -61,13 +61,39 @@ class CTxIn
 
+    /* Setting nSequence to this value for every input in a transaction
+     * disables nLockTime. */
+    static const uint32_t SEQUENCE_FINAL = 0xffffffff;
```
The comments here are slightly confusing to me because of missing context.  The above is a consensus rule for all transactions, while the ones below only apply as policy (for now) for `tx.nVersion >= 2` transactions.  I think we should expand on the comments to make this clearer, perhaps with a reference to BIP68?

btcdrak replied:
Agreed. I think it's pretty sure this code or #7187 is going supersede #6312 so I'm fine with the comment being updated here. 

maybe

    /* Setting nSequence to this value for every input in a transaction
     * disables nLockTime when transaction nVersion is >=2 . */

jtimon replied:
It's not clear to me that this is is going to supersede #6312 . It seems there's at least 2 more people with alternatives. Can we maintain #6312 updated as a common base for the 3 different options?
As said, squashes can happe right before merging (you can also reset HEAD^ the squashed commit and redo it with your name if that's the reason why this currently doesn't use the common base [no other reason comes to mind]).

maaku replied:
This comment has nothing to do with BIP 68. Setting all nSequence to max into disables nLockTime for all tx versions, today.

sdaftuar commented on src/main.cpp:
```diff
@@ -705,6 +706,116 @@ bool CheckFinalTx(const CTransaction &tx, int flags)
+
+static bool EvaluateSequenceLocks(const CBlockIndex& block, std::pair<int, int64_t> lockPair)
+{
+    int64_t nBlockTime =  block.GetAncestor(std::max(block.nHeight-1, 0))->GetMedianTimePast();
```
I think using `block.nHeight-1` here forces an unnecessary restriction on the caller (ie that `block.nHeight` is being set to 1 more than the current tip height -- if we change this to be 
```
int64_t nBlockTime = block.pprev ? block.pprev->GetMedianTimePast() : 0;
```
Then we would be able to change the `CheckSequenceLocks` code in the future to allow us to test whether a transaction is able to be confirmed in the next N blocks for N > 1 (right now we can only do N=1).  

This would allow for mempool policy code that could permit transactions that may not be mineable now, but will be mineable in the near future.

NicolasDorier replied:
The semantic used by caller is the following : 
*Check if SequenceLock knowing the transaction is included in "block"*

This is needed. Because *block* might be an existing old block but also a block on top of tip. (for mempool transactions)
Changing the semantic of this method would make things less clear.

I introduced this change on https://github.com/maaku/bitcoin/commit/aa83819cac79edc3ab9c140cbbd2cd1417e04a8a . Previously, the LockTime method needed both : the tip and the block where it was tested against. I simplified by making it requires only the block it is tested against.

sdaftuar replied:
@NicolasDorier My point is that we're requiring `nHeight-1` to be a valid blockindex entry, when really all we need to do is rely on `pprev` being a valid block index (modulo checks for the genesis block).

If we use `pprev` for the time calculation rather than `nHeight`, that would give us the freedom to tweak the value of `nHeight` that we pass in, so that we could test for transactions with height-based-relative-locks being confirmable in N blocks rather than just in the next block.

NicolasDorier replied:
woops, sorry I misunderstood your point. (I thought you wanted to pass the previous block directly) 
Correct.

rustyrussell commented:
Tested-by: Rusty Russell <rusty@rustcorp.com.au> (merged with #6564 for testing, with only trivial header conflict fixup).

morcos commented:
@sdaftuar's nits and suggestions addressed.

After this receives some ACK's I can work on back ports

NicolasDorier commented:
I think a new PR with #7187 on top of this one should be open. It makes no sense to review only this PR without #7187.

jtimon commented:
Agree with @NicolasDorier. If it wasn't  for #7187 there would be no reason to replace #6312 in the first place.

btcdrak commented:
Test script is available here (requires merge of #6564 with #7184) https://github.com/ajtowns/op_csv-test/blob/master/test_csv_htlc.py

petertodd commented on src/main.cpp:
```diff
@@ -705,6 +706,122 @@ bool CheckFinalTx(const CTransaction &tx, int flags)
+
+static bool EvaluateSequenceLocks(const CBlockIndex& block, std::pair<int, int64_t> lockPair)
+{
+    int64_t nBlockTime = block.pprev ? block.pprev->GetMedianTimePast() : 0;
```
Would you mind adding a comment explaining when block.pprev might be null; if I understand it correctly the only case where that is true is the genesis block, in which case it'd be good to add an assertion that the block height matches that case.

morcos replied:
I'm not sure why we think its ever necessary to check sequence locks on a genesis block.
So what would you think about assert(block.pprev) and then removing the ternary operator?

btcdrak replied:
@morcos This is the reason for the ternary https://github.com/bitcoin/bitcoin/pull/6312#discussion_r44622250

morcos replied:
@btcdrak That was the old code.  The new code calls SequenceLocks from ConnectBlock which doesn't run on the genesis block and doesn't call SequenceLocks on coinbase txs.  If both of those things change, it would get caught by the proposed assert.

jtimon replied:
Ack on assert(block.pprev): we should never validate the genesis block anyway  (it's not only correct, it is the first consensus rule!).

petertodd commented on src/main.cpp:
```diff
@@ -705,6 +706,122 @@ bool CheckFinalTx(const CTransaction &tx, int flags)
+        const CTxIn& txin = tx.vin[txinIndex];
+        CCoins coins;
+        if (!viewMemPool.GetCoins(txin.prevout.hash, coins)) {
+            return  error("%s: Missing input", __func__);
```
nit: extra space after return

petertodd commented:
ACK https://github.com/morcos/bitcoin/commit/e87d06a20e7d1dd9f3000c419a4dfe805e3dbec4, modulo nits fixed in https://github.com/morcos/bitcoin/pull/7

That said, while I'm happy to see this merged as a mempool-only thing, before we actually soft-fork this in I'd like to see more unit tests; it's notable and worrying that I could change  https://github.com/morcos/bitcoin/blob/e87d06a20e7d1dd9f3000c419a4dfe805e3dbec4/src/main.cpp#L760 to remove the -1 and only one test failed. It might be good to extend the format of the transaction_(in)valid.json tests to let you specify txin heights/blocktimes, and then create some dummy block and coin indexes to use it. Another option might be to make a data-driven block acceptance unittest.

NicolasDorier commented:
@petertodd can you review #7187 as well ? Those two PR goes together.

CodeShark commented:
@morcos I just gave this PR a once over - very clean implementation! I'll run a few more tests.

morcos commented:
@petertodd I addressed your nits, but did it a bit differently than your suggestions.  Let me know if you're ok with this.

In particular though, it occurs to me that the subtracting 1 to retain nLockTime semantics is an artifact of how this code used to be combined with nLockTime checks.  These semantics are purely internal.  I think it might be cleaner to not subtract 1 and change the comparison to be a  strict >?  It's not clear to me what will be more confusing for later coders, because we will be saving these effective lock times in LockPoints as in #7187, so maybe keeping the nLockTime semantics makes sense?

maaku commented:
Keeping the existing semantics makes a lot more sense, as silly as they
are, especially if there is any future refactorings (unknown at this time)
that generalize nLockTime and nSequence behavior.

On Mon, Feb 1, 2016 at 1:26 PM, Alex Morcos <notifications@github.com>
wrote:

> @petertodd <https://github.com/petertodd> I addressed your nits, but did
> it a bit differently than your suggestions. Let me know if you're ok with
> this.
>
> In particular though, it occurs to me that the subtracting 1 to retain
> nLockTime semantics is an artifact of how this code used to be combined
> with nLockTime checks. These semantics are purely internal. I think it
> might be cleaner to not subtract 1 and change the comparison to be a strict
> >? It's not clear to me what will be more confusing for later coders,
> because we will be saving these effective lock times in LockPoints as in
> #7187 <https://github.com/bitcoin/bitcoin/pull/7187>, so maybe keeping
> the nLockTime semantics makes sense?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/7184#issuecomment-178201104>.
>

josephpoon commented:
Tested ACK using btcdrak's BIP68+OP_CSV combined branch https://github.com/bitcoin/bitcoin/compare/master...btcdrak:sequenceandcsv on regtest.

This pull request is _very_ useful for Lightning Network channels without pre-set expiries. Thanks~~~!

CodeShark commented:
Tested ACK on regtest

petertodd commented:
utACK https://github.com/morcos/bitcoin/commit/ca786956afa9bf344255959447b0386dd142ebd1

Roasbeef commented:
LGTM. Err, I mean, Tested ACK :) 

Testing methodology: 
  * Pulled down @btcdrak's combined  BIP68+OP_CSV branch: https://github.com/bitcoin/bitcoin/compare/master...btcdrak:sequenceandcsv. 
  * Started a local Bitcoin Core node with the `-regtest` flag activated. 
  * I then created an [integration test](https://github.com/LightningNetwork/lnd/commit/05f4d1fa9721f7e8c44844b719294a5f06cc21a9) in a [new branch](https://github.com/LightningNetwork/lnd/tree/csv-test) within [lnd](https://github.com/LightningNetwork/lnd) created specifically to test the combined BIP 68+112 branch. This particular test case directly exercises a possible scenario within LN, wherein one of the sides immediately broadcasts the current (and only) commitment transaction. As a result of this uncooperative closure, the party that broadcasted the transaction must wait a CSV delay before they can re-claim their funds.
  * The test case exercises the following negative paths when verifying a transaction attempting to spend a CSV output, additionally ensuring the txn spending the CSV output is mature enough: 
     * invalid tx version
     * mismatched sequence numbers
     * sequence disable flag set
     * CSV delay pre-mature 

The test case detailed above can be found [here](https://github.com/LightningNetwork/lnd/blob/csv-test/lnwallet/script_utils_test.go). 

A major thanks to all those involved from BIP drafting, to BIP review, to coding the reference implementation, to code review, and finally, testing. As Joseph said above, relative time-locks are _extremely_ useful within the Lightning Network 😎.

NicolasDorier commented:
@Roasbeef @josephpoon please, can you test this PR as well. @btcdrak branch has been supersede and this one is relatively different. (but should achieve the same purpose)
Once SW and that one get pushed, I'll jump on board on Lightning Network with the others cool boys ! :+1:

btcdrak commented:
@NicolasDorier I think you are confused. The combined branch _is this PR_.

NicolasDorier commented:
Ok my bad, I tought your PR was #6312

CodeShark commented:
Tested ACK using btcdrak's BIP68+OP_CSV combined branch master...btcdrak:sequenceandcsv on regtest.

instagibbs commented on src/main.h:
```diff
@@ -341,7 +341,22 @@ bool IsFinalTx(const CTransaction &tx, int nBlockHeight, int64_t nBlockTime);
 
-/** 
+/**
+ * Check if transaction is final per BIP 68 sequence numbers and can be included in a block.
```
nit: s/in a block/in a given block/

Took me a few readings to figure out what it was saying.

instagibbs commented:
utACK

NicolasDorier commented on src/main.cpp:
```diff
@@ -705,6 +706,127 @@ bool CheckFinalTx(const CTransaction &tx, int flags)
+static bool EvaluateSequenceLocks(const CBlockIndex& block, std::pair<int, int64_t> lockPair)
+{
+    assert(block.pprev);
+    int64_t nBlockTime = block.pprev->GetMedianTimePast();
```
What if block is genesis ?

jtimon replied:
If it's the genesis block we know is valid and we don't need to check it. Hasn't this been discussed in an outdated diff already?

morcos replied:
See https://github.com/bitcoin/bitcoin/pull/7184#discussion_r51482569

NicolasDorier commented:
Tested ACK, reimplemented  CSV from https://github.com/bitcoin/bitcoin/compare/master...btcdrak:sequenceandcsv in NBitcoin and did some tests at transaction level. (https://github.com/MetacoSA/NBitcoin/blob/csv/NBitcoin.Tests/transaction_tests.cs#L862)

sipa commented:
A (harmless) red flag: when CalculateSequenceLocks is called from CheckSequenceLocks, its passed CBlockIndex& object is not fully initialized (only pprev and nHeight are set), so calling GetAncestor looked dangerous to me. It seems that GetAncestor deals correctly with this case, however, and will fall back to walking using pprev instead of pskip in that case.

sipa commented on src/main.cpp:
```diff
@@ -705,6 +706,127 @@ bool CheckFinalTx(const CTransaction &tx, int flags)
+
+bool CheckSequenceLocks(const CTransaction &tx, int flags)
+{
+    AssertLockHeld(cs_main);
```
Also needs a lock on the mempool, I think.

morcos replied:
added

sipa commented on src/main.cpp:
```diff
@@ -705,6 +706,127 @@ bool CheckFinalTx(const CTransaction &tx, int flags)
+    // evaluated is what is used. Thus if we want to know if a
+    // transaction can be part of the *next* block, we need to call
+    // SequenceLocks() with one more than chainActive.Height().
+    index.nHeight = tip->nHeight + 1;
```
It's also needed because of internal consistency. If x->pprev->nHeight is X, x->nHeight must be X+1.

morcos replied:
Yep.  Originally the intention had been to be able to set this to a higher number if we wanted to allow the mempool to contain currently locked transactions.  I discovered that that doesn't work because of the GetAncestor use in CalculateSequenceLocks.  I wrote  up some minor changes to pass the height/time you want to check against directly to EvaluateSequenceLocks, but I think it's not worth making those changes now, we can do it if/when we decide we want to keep non-final txs.

sipa replied:
Yeah, not worth changing things for. Just saying that there isn't really any other option.

sipa commented:
utACK

sipa commented on src/main.cpp:
```diff
@@ -951,6 +1073,12 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState &state, const C
         view.SetBackend(dummy);
         }
 
+        // Only accept BIP68 sequence locked transactions that can be mined in the next
```
This should be moved up to be within the above mempool-locked block.

morcos commented:
Oops.  Nice catch @sipa.   Fixed the missing lock.  It's sort of a shame not to reuse the CCoinsViewCache just created in ATMP, but we can save that potential improvement for another time.

btcdrak commented:
> Nice catch @sipa.

@sipa is a rockstar! :guitar:

sipa commented:
test/miner_tests.cpp:82

unknown location(0): fatal error: in "miner_tests/CreateNewBlock_validity": signal: SIGABRT (application abort requested)

sipa commented:
utACK 4315fe1afce448698732cf1bcffeb070028e4b2b after squashing 

Or, utACK tree id 04ef32caaaf2b838295f39cffbdd87468f3657c9

```
$ git show -s --format="%T" 4315fe1afce448698732cf1bcffeb070028e4b2b
04ef32caaaf2b838295f39cffbdd87468f3657c9
```

A squash should retain the tree id.

morcos commented:
Ok squashed, same code 
```
git show -s --format="%T" da6ad5f684b91975cae3f37495ccbd041499e86b
04ef32caaaf2b838295f39cffbdd87468f3657c9
```

morcos committed:
Bug fix to RPC test

morcos commented:
sigh..  small rarely occurring bug in the RPC test
@laanwj when you are ready to merge, i'll squash again, or feel free to just it yourself

maaku commented:
Meta-point, it would be better process to not squash these PRs...

sipa commented:
@maaku My view is that the commit history of what is merged should optimize for reviewability by someone who has not seen the pull requests. So multiple commits that implement separate parts of a feature, make sense on their own, and result in a working codebase on their own are certainly welcome, but I don't think we should aim for having the exact history of a commit, and certainly not when some commits are broken and need follow-up commits for fixing.

laanwj commented:
> @laanwj when you are ready to merge, i'll squash again, or feel free to just do it yourself

Okay, thanks.

> Meta-point, it would be better process to not squash these PRs...

I don't agree on that. One should try to order commits so that it does not break atomicity, or create intermediate pulls that fail buld/test, as that messes with `git bisect`.
Also @sipa makes a good point that commit history should be optimized for reviewing, not what happened to be the chronological order of changes at the time. This may involve multiple commits if that makes reviewing easier (due to separate concerns, for example, or moves+changes).

Putting the commit hash in comments is a help to compare the version that was reviewed to the version that was merged - it is not meant to reduce flexibility of the developer with regard to ordering commits.

Process discussion is completely off-topic here though, so let's leave it at that.

sdaftuar commented on src/main.cpp:
```diff
@@ -705,6 +706,127 @@ bool CheckFinalTx(const CTransaction &tx, int flags)
+    index.pprev = tip;
+    // CheckSequenceLocks() uses chainActive.Height()+1 to evaluate
+    // height based locks because when SequenceLocks() is called within
+    // CBlock::AcceptBlock(), the height of the block *being*
```
nit: `CBlock::AcceptBlock()` isn't a thing, should say `ConnectBlock` I think.

sdaftuar commented on src/main.cpp:
```diff
@@ -705,6 +706,127 @@ bool CheckFinalTx(const CTransaction &tx, int flags)
+    // CBlock::AcceptBlock(), the height of the block *being*
+    // evaluated is what is used. Thus if we want to know if a
+    // transaction can be part of the *next* block, we need to call
+    // SequenceLocks() with one more than chainActive.Height().
```
nit: This sentence is somewhat unclear, as `SequenceLocks` isn't invoked here.

sdaftuar commented on src/main.cpp:
```diff
@@ -2098,6 +2228,19 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
+            }
+
+            if (!SequenceLocks(tx, nLockTimeFlags, &prevheights, *pindex)) {
+                return state.DoS(100, error("ConnectBlock(): contains a non-BIP68-final transaction", __func__),
```
Need to add a `%s` or drop this `__func__`.

sdaftuar commented on src/main.h:
```diff
@@ -341,7 +341,22 @@ bool IsFinalTx(const CTransaction &tx, int nBlockHeight, int64_t nBlockTime);
+/**
+ * Check if transaction will be BIP 68 final in the next block to be created.
+ *
+ * Calls SequenceLocks() with data from the tip of the current active chain.
```
This comment is incorrect (`CheckSequenceLocks` doesn't actually call `SequenceLocks`).

sdaftuar commented on src/policy/policy.h:
```diff
@@ -45,8 +45,9 @@ static const unsigned int STANDARD_SCRIPT_VERIFY_FLAGS = MANDATORY_SCRIPT_VERIFY
 
-/** Used as the flags parameter to CheckFinalTx() in non-consensus code */
-static const unsigned int STANDARD_LOCKTIME_VERIFY_FLAGS = LOCKTIME_MEDIAN_TIME_PAST;
+/** Used as the flags parameter to LockTime() in non-consensus code. */
```
s/`LockTime`/`CheckSequenceLocks`/

sipa commented:
Mental note: verify how this interacts with the wallet

morcos committed:
fix sdaftuar's nits again

it boggles the mind why these nits can't be delivered on a more timely basis

sdaftuar commented:
ACK b043c4b746c8199ce948aa5e8b186e0d1a61ad68

laanwj commented:
utACK b043c4b
